### PR TITLE
[FRONTEND] A python-tvm hybrid programming model

### PR DESCRIPTION
**Pull request by were**
This hybrid frontend is aimed at:
1. Building IR in a more intuitive way
2. Writing preliminary versions of some idioms that yet have not been supported by

# Features

## Software emulation
This feature supports both software emulation and compilation of the code.

To define a function, you need to use `tvm.hybrid.script` decorator to indicate this is a hybrid function:
````Python
@tvm.hybrid.script
def outer_product(a, b, c):
    for i in range(a.shape[0]):
        for j in range(b.shape[0]):
            c[i, j] = a[i] * b[j]
a = numpy.random.rand(100)
b = numpy.random.rand(99)
c = numpy.zeros((100, 99))
outer_product(a, b)
````
This decorator will help you to import [key words](#keywords) required spontaneously when software emulation.
Every element in the argument list is either a python variable or `numpy` tensor.
## Backend Compilation
The current parse interface looks like:
````Python
a = tvm.placeholder((100, ), name='a')
b = tvm.placeholder((99, ), name='b')
c = tvm.placeholder((100, 99), name='c')
tvm.hybrid.parse(outer_product, [a, b, c]) # return an ir root of this function
````
**TODO**: If we pass these tvm tensors to this function, it returns a op node:
````Python
a = tvm.placeholder((100, ), name='a')
b = tvm.placeholder((99, ), name='b')
c = tvm.placeholder((100, 99), name='c')
op = outer_product(a, b, c) # return the corresponding op node
````
### Scheduling
**Under construction, not truly supported yet.**

Follow up the example above, you can use some tvm like interfaces to manipulate the structure of IR:
````Python
sch = tvm.create_schedule(op)
jo, ji = sch.split(j, 4)
sch.vectorize(ji)
````
`split`, `reorder`, and loop_annotation will be supported!
## Attributes
So far, ONLY tensors' `shape` attribute is supported!

## Loops
In HalideIR, loops have in total 4 types: `serail`, `unrolled`, `parallel`, and `vectorized`.

Here we use `range`, `serial`, `unroll`, `parallel`, and `vectorize`, these **5** keywords to annotate the types of for loops.

**NOTE**: In HalideIR those are enums, they are in passive form. Here we use active form to annotate loops, because they are ready to run.

**NOTE**: Unlike what that is in HalideIR, in `loop_type(a, b)`, `a` is the starting point and `b` is the trip count of iterations. Here `loop_type(a, b)` indicates `[a, b)`.

## Variables
Because there is no variables in `HalideIR`, all the mutatable variables will be lowered to an array with size 1.
It takes the first store of a variable as its declaration.
**NOTE**: Unlike conventional Python, the declared array can only be used in the scope level it is declared.
````Python
for i in range(5):
    sum = 0
    for j in range(5):
    	sum += a[i, j] #do something with sum
    b[i] = sum #you can still use sum in this level
#you can NEVER use some here, even though it is allowed in conventional Python
a[0] = sum
````
## Conditional Statement and Expression
````Python
if condition:
    # do something
a = b if condition else c
````
However, NO `True` and `False` keyword supported yet.
## Math intrinsics
So far, these math intrinsics, `log`, `exp`, `sigmoid`, `tanh`, `power`, and `popcount`, are supported. No import is required, just use it!
## Array allocation
**TODO**: Use a function call `allocation(shape, type, share/local)` to declare an array buffer. The basic usage is roughly the same as variables
## Thread bind
You can also do loop-thread bind by writing code like this:
````Python
for tx in bind("threadIdx.x", 100):
    a[tx] = b[tx]
````
# Appendix

## <a name="keywords"> Keywords </a>
- Statement keywords: `for`, `in`, `if`, `else`
- For keywords: `serial`, `range`, `unroll`, `parallel`, `vectorize`, `bind`
- Math keywords: `log`, `exp`, `sigmoid`, `tanh`, `power`, `popcount`

### CODE REVIEW

**Comment by tqchen** [comment]
@kevinthesun @ZihengJiang @merrymercy can you help review this?

**Comment by tqchen** [comment]
Please add document about the language, e.g. we could use some decorator like 
```
@tvm.ir_script
```

**Comment by tqchen** [comment]
this could due to python2/python3 compact issue, try to remove the print(which shouldn't be in package anyway)

**Comment by were** [comment]
- Interface: `tvm.contrib.pyfrontend.parse(func, args)` and `tvm.contrib.pyfrontend.lower(func, args, binds)`
    - `func` is either a string or a python **GLOBAL** function. If it is a subfunction in a function or a method in a class, an error will occur.
    - `args` is the argument list of this function.
    - `binds` the same as the `binds` argument in `tvm.lower`
    - `parse` is something like `simple_mode=True` in `tvm.lower`
    - `tvm.contrib.pyfrontend.lower` is something like `simple_mode=False` in `tvm.lower`
    - The IR generated by `lower` is compatible with tvm.build, so we can get LLVM JIT easily

- Syntax supported:
    - **Main arithmetic ops: ** four arithmetic ops and bitwise ops are supported. Math functions have not been implemented.
    - **Loops: ** only two kinds of loops are supported (the loop bound can be variable but the step **MUST** be 1):
````python
for i in range(5):
for i in range(0, n):
````
    - **Variables: ** because there is no mutable variables in HalideIR, all the variables will be replaced by a buffer with size 1.
    - **Annotation: ** we can use `with` to annotate loops, but we need to add decorator, `tvm.contrib.pyfrontend.py_fronted`, before the functions to ensure the sematic; I am not sure if it is a good idea: the decorator forcely put the class `Unrolled` into the global scope

- Summary
To sum up, code that uses `pyfrontend` should roughly look like:
````python
@tvm.contrib.pyfronted.py_frontend
def foo(a, b):
    sum = 0.0
    with Unrolled as i:
        for i in range(5):
            sum += a[i] * a[i]
    b[0] = a[i]

a = tvm.placeholder((5, ), dtype='float32', name='a')
b = tvm.placeholder((1, ), dtype='float32', name='b')
tvm.contrib.pyfrontend.parse(foo, [a, b])
````

File: python/tvm/contrib/pyfrontend.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+    ----------
+    func : str or types.FunctionType
+        If it is a string, parse the source code.
+        If it is a function, parse the function.
+
+    args : list of Buffer or Tensor or Var
+        The argument lists to the function.
+        Leave it None if no buffer is related to the function to be parsed
+
+    dump : bool, optional
```
Comment (xqdan) [inline python/tvm/contrib/pyfrontend.py]: You may add it in parse above or just remove description here.

File: python/tvm/contrib/pyfrontend.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+from .. import stmt as _stmt
+from .. import make as _make
+from .. import api  as _api
+from .. import ir_pass  as _ir_pass
+from .. import build_module as builder
+
+#pylint: disable=missing-docstring, invalid-name, consider-merging-isinstance, no-else-return
+#pylint: disable=inconsistent-return-statements, eval-used
+
+NOOP = _make.Evaluate(_api.const(0))
```
Comment (xqdan) [inline python/tvm/contrib/pyfrontend.py]: We may use NOP, the same meaning with NOP in some hardware instruction sets, 
eg http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0489h/Cjafcggi.html.
```
NOP
No Operation.
Show/hideSyntax
NOP{cond}
```

File: python/tvm/contrib/pyfrontend.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+class PyAST2HalideIR(ast.NodeVisitor):
+
+    _binop_maker = {
+        ast.Add   : _make.Add,
+        ast.Sub   : _make.Sub,
+        ast.Mult  : _make.Mul,
+        ast.Div   : _make.Div,
+        ast.Mod   : _make.Mod,
+        ast.BitOr : operator.or_,
```
Comment (xqdan) [inline python/tvm/contrib/pyfrontend.py]: A question here,  we have all operations above in halide, why we have two different ir building here?

File: tests/python/unittest/test_halide_text_format.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+    tvm_a = tvm.ndarray.array(_a)
+    tvm_b = tvm.ndarray.array(_b)
+    tvm_c = tvm.ndarray.array(numpy.zeros((_n, _m), dtype = 'float32'))
+    func(_n, _m, tvm_a, tvm_b, tvm_c)
+    numpy.testing.assert_allclose(tvm_c.asnumpy(), c_python, rtol = 1e-5)
+    #print(func)
+
+def fanout(n, a, b):
+    three = 3.0
```
Comment (xqdan) [inline tests/python/unittest/test_halide_text_format.py]: can we support while?

File: tests/python/unittest/test_halide_text_format.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+def test_failure():
+    try:
+        frontend.parse(failure, [])
+    except AssertionError:
+        _, _, tb = sys.exc_info()
+        _, _, func, text = traceback.extract_tb(tb)[-1]
+        assert func == 'visit_Name'
+        assert text == 'assert node.id not in self.loops_above'
+
+@tvm.contrib.pyfrontend.py_frontend
```
Comment (xqdan) [inline tests/python/unittest/test_halide_text_format.py]: IMO, we may put this frontend in tvm/python instead of contrib.

File: python/tvm/contrib/pyfrontend.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
+"""Compiling a subset of Python to HalideIR"""
+import ast
```
Comment (xqdan) [inline python/tvm/contrib/pyfrontend.py]: Good to learn: Ast is a python package, which is similar with Ast in halide. You may add a comment for code readers.

File: tests/python/unittest/test_halide_text_format.py (hunk 5 above)
Comment (xqdan) [reply tests/python/unittest/test_halide_text_format.py]: @tqchen

File: python/tvm/contrib/pyfrontend.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    var_usage = _find_all_variable_decl(ir)
+    parser = PyAST2HalideIR(args, var_usage)
+    stmt = parser.visit(ir)
+    binds_args, args = builder.get_binds(parser.args)
+    binds_vars, _ = builder.get_binds(parser.vars_buffer)
+    _binds = binds_args.copy()
+    _binds.update(binds_vars)
+    _binds.update(binds_vars)
+    _binds.update(binds)
+    stmt = _ir_pass.StorageFlatten(stmt, _binds, 64)
```
Comment (xqdan) [inline python/tvm/contrib/pyfrontend.py]: did you try other passes?

File: python/tvm/contrib/pyfrontend.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+    if binds is None:
+        binds = {}
+    var_usage = _find_all_variable_decl(ir)
+    parser = PyAST2HalideIR(args, var_usage)
+    stmt = parser.visit(ir)
+    binds_args, args = builder.get_binds(parser.args)
+    binds_vars, _ = builder.get_binds(parser.vars_buffer)
+    _binds = binds_args.copy()
+    _binds.update(binds_vars)
+    _binds.update(binds_vars)
```
Comment (merrymercy) [inline python/tvm/contrib/pyfrontend.py]: duplicated function call?

File: tests/python/unittest/test_halide_text_format.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    assert len(value.a.args) == 1
+    assert value.a.args[0].value == 0
+    assert abs(value.b.value - (1 / 3.0)) < 1e-5
+    write = ibody.rest.rest.rest
+    assert isinstance(write, tvm.stmt.Provide)
+    assert write.func.name == 'b'
+    assert write.value.name == 'sigma'
+    assert len(write.value.args) == 1
+    assert write.value.args[0].value == 0
+    #print(write)
```
Comment (merrymercy) [inline tests/python/unittest/test_halide_text_format.py]: Add correctness test with tvm.build, just like the above unit test.

File: python/tvm/contrib/pyfrontend.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+        assert isinstance(node.target, ast.Name)
+        var = node.target.id
+        self.loop_levels[var] = _api.var(var)
+        assert isinstance(node.iter, ast.Call)
+        assert node.iter.func.id == 'range'
+        assert len(node.iter.args) == 1 or len(node.iter.args) == 2
+        if len(node.iter.args) == 1:
+            low, high = _api.const(0), self.visit(node.iter.args[0])
+        elif len(node.iter.args) == 2:
+            low, high = self.visit(node.iter.args[0]), self.visit(node.iter.args[1])
```
Comment (merrymercy) [inline python/tvm/contrib/pyfrontend.py]: Can we support triangle access now?

File: tests/python/unittest/test_halide_text_format.py (hunk 4 above)
Comment (were) [reply tests/python/unittest/test_halide_text_format.py]: No, because no `While` node in `HalideIR`

File: python/tvm/contrib/pyfrontend.py (hunk 1 above)
Comment (were) [reply python/tvm/contrib/pyfrontend.py]: OK. Let me do this.

File: tests/python/unittest/test_halide_text_format.py (hunk 5 above)
Comment (were) [reply tests/python/unittest/test_halide_text_format.py]: Maybe it is a good idea, but Tianqi told me that we should put immature functionality in `contrib` first.

File: tests/python/unittest/test_halide_text_format.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        sigma = 0.0
+        for j in range(3):
+            sigma = sigma + a[i + j]
+        sigma = sigma / three
+        b[i] = sigma
+
+def test_fanout():
+    n = tvm.var('n')
+    a = tvm.placeholder((n, ), name = 'a')
+    b = tvm.placeholder((n-3, ), name = 'b')
```
[the remark this answers is not in the document]
Comment (were) [reply tests/python/unittest/test_halide_text_format.py]: One of my original purposes is to both replace `ir_builder`. If one day `ir_builder` is replaced this should be all refactored.

File: python/tvm/contrib/pyfrontend.py (hunk 6 above)
Comment (were) [reply python/tvm/contrib/pyfrontend.py]: Actually, this is not the cardinal problem. The cardinal problem I am suffering from is that `ast` in py2 and py3 are quite different which makes me not pass the test yet.

File: python/tvm/contrib/pyfrontend.py (hunk 7 above)
Comment (were) [reply python/tvm/contrib/pyfrontend.py]: No. I only tried this pass so far. Later, when more functionalities are supported, I will put more passes. So far, this pass is already sufficient to glue LLVM JIT backend.

File: tests/python/unittest/test_halide_text_format.py (hunk 9 above)
Comment (were) [reply tests/python/unittest/test_halide_text_format.py]: I did! In first, I have a function both run in numpy and module built by my pass.

File: python/tvm/contrib/pyfrontend.py (hunk 8 above)
Comment (were) [reply python/tvm/contrib/pyfrontend.py]: My bad. I will fix it!

File: python/tvm/contrib/pyfrontend.py [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+
+    def visit_Subscript(self, node):
+        #assert isinstance(node.value, ast.Name) or isinstance(node.value, ast.Attribute)
+        args = self.visit(node.slice)
+        if isinstance(node.value, ast.Name):
+            array = node.value.id
+            assert array in self.input_param.keys()
+            _buf = self.input_param[array]
+            return _make.Call(_buf.dtype, array, args, _expr.Call.Halide, _buf.op, 0)
+        elif isinstance(node.value, ast.Attribute):
```
[the remark this answers is not in the document]
Comment (were) [reply python/tvm/contrib/pyfrontend.py]: OK. I add one.

File: python/tvm/contrib/pyfrontend.py (hunk 10 above)
Comment (were) [reply python/tvm/contrib/pyfrontend.py]: I think so.

File: tests/python/unittest/test_halide_text_format.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+    _a = numpy.random.rand(_n).astype('float32')
+    _b = numpy.random.rand(_m).astype('float32')
+    c_python = numpy.zeros((_n, _m), dtype = 'float32')
+    outer_product(_n, _m, _a, _b, c_python)
+
+    tvm_a = tvm.ndarray.array(_a)
+    tvm_b = tvm.ndarray.array(_b)
+    tvm_c = tvm.ndarray.array(numpy.zeros((_n, _m), dtype = 'float32'))
+    func(_n, _m, tvm_a, tvm_b, tvm_c)
+    numpy.testing.assert_allclose(tvm_c.asnumpy(), c_python, rtol = 1e-5)
```
Comment (were) [inline tests/python/unittest/test_halide_text_format.py]: @merrymercy 
Here is the comparison between numpy and code built by my pass.

File: python/tvm/contrib/pyfrontend.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (were) [reply python/tvm/contrib/pyfrontend.py]: Added.

File: tests/python/unittest/test_halide_text_format.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (xqdan) [reply tests/python/unittest/test_halide_text_format.py]: That's an issue need to discuss. lots of code/backend is based on it, especially emiting intrinsics for targets, so IMO, ir_builder could be repalced in new code, but we should keep it in mainstream.

File: python/tvm/contrib/pyfrontend.py (hunk 3 above)
Comment (were) [reply python/tvm/contrib/pyfrontend.py]: This is only for parsing and building IR node.

File: python/tvm/contrib/pyfrontend.py (hunk 2 above)
Comment (were) [reply python/tvm/contrib/pyfrontend.py]: This is done.

**Comment by tqchen** [comment]
I posted some of my followup thoughts in the RFC https://discuss.tvmlang.org/t/text-format-for-halide-ir-from-a-subset-of-python/103

**Comment by kevinthesun** [comment]
Will this frontend support binding threads. Currently it's not very easy to do such task with ir builder.

File: tests/python/unittest/test_halide_text_format.py [hunk 14]
```diff
@@ -127,13 +127,18 @@ def fanout(n, a, b):
     assert len(write.value.args) == 1
     assert write.value.args[0].value == 0
 
+@hybrid_script
```
Comment (tqchen) [inline tests/python/unittest/test_halide_text_format.py]: rename to test_hybrid_script.py

File: docs/dev/hybrid_script.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
+This hybrid frontend is aimed at:
```
Comment (tqchen) [inline docs/dev/hybrid_script.md]: Starts with a title

File: docs/dev/hybrid_script.md [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+You can also do loop-thread bind by writing code like this:
+````Python
+for tx in bind("threadIdx.x", 100):
+    a[tx] = b[tx]
+````
+# Appendix
+
+## <a name="keywords"> Keywords </a>
+- Statement keywords: `for`, `in`, `if`, `else`
+- For keywords: `serial`, `range`, `unroll`, `parallel`, `vectorize`, `bind`
```
Comment (tqchen) [inline docs/dev/hybrid_script.md]: let us remove serial(as range defaults to serial) and use the begin end convention for all of our loop constructs

**Comment by were** [comment]
OK. I close this PR and open a new one to clean up the code.
